oauth2_http/java/com/google/auth/oauth2/LoggingUtils.java

```diff
@@ -0,0 +1,207 @@
+
+package com.google.auth.oauth2;
+
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.http.UrlEncodedContent;
+import com.google.api.client.http.json.JsonHttpContent;
+import com.google.api.client.util.GenericData;
+import com.google.gson.Gson;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.MDC;
+
+class LoggingUtils {
+
+  private static final Gson gson = new Gson();
+  private static final Set<String> sensitiveKeys =
+      new HashSet<>(
+          Arrays.asList(
+              "token",
+              "assertion",
+              "access_token",
+              "client_secret",
+              "refresh_token",
+              "signedBlob"));
+
+  private LoggingUtils() {}
+
+  static void logWithMDC(
+      Logger logger, org.slf4j.event.Level level, Map<String, String> contextMap, String message) {
+    if (!contextMap.isEmpty()) {
+      contextMap.forEach(MDC::put);
+      contextMap.put("message", message);
+      message = gson.toJson(contextMap);
+    }
+    switch (level) {
+      case TRACE:
+        logger.trace(message);
+        break;
+      case DEBUG:
+        logger.debug(message);
+        break;
+      case INFO:
+        logger.info(message);
+        break;
+      case WARN:
+        logger.warn(message);
+        break;
+      case ERROR:
+        logger.error(message);
+        break;
+      default:
+        logger.info(message);
+        // Default to INFO level
+    }
+    if (!contextMap.isEmpty()) {
+      MDC.clear();
+    }
+  }
+
+  static void logRequest(HttpRequest request, Logger logger, String message) {
+    try {
+      if (logger.isInfoEnabled()) {
+        Map<String, String> loggingDataMap = new HashMap<>();
+        loggingDataMap.put("request.method", request.getRequestMethod());
+        loggingDataMap.put("request.url", request.getUrl().toString());
+
+        Map<String, Object> headers = new HashMap<>();
+        request
+            .getHeaders()
+            .forEach(
+                (key, val) -> {
+                  if ("authorization".equals(key)) {
+                    String hashedVal = calculateSHA256Hash(String.valueOf(val));
+                    headers.put(key, hashedVal);
+                  } else {
+                    headers.put(key, val);
+                  }
+                });
+        loggingDataMap.put("request.headers", headers.toString());
+
+        if (request.getContent() != null && logger.isDebugEnabled()) {
+          // are payload always GenericData? If so, can parse and store in json
+          if (request.getContent() instanceof UrlEncodedContent) {
+
+            GenericData data = (GenericData) ((UrlEncodedContent) request.getContent()).getData();
```

Comment:
Can you add a quick comment about why UrlEncodedContent is casted to GenericData? But JsonHttpContent (below) doesn't need to be?

Reply:
added